### DISCUSSION TITLE
Make semantics of buffer/window numbers/IDs more consistent

### DISCUSSION THREAD

**Discussion by adigitoleo** [neovim/neovim#30073] [Ideas]
After messing around with floatwin api for a while in various attempts to make plugins, I've been bitten repeatedly by the unpredictability and complexity of the semantics of window/buffer numbers.

The most uncomfortable parts are the use of `0` and `-1` which can mean various things in different contexts, which is not always well documented.

**For buffers:**
1. `nvim_create_buf` returns `0` on failure
2. `bufnr` accepts the `{buf}` arg, whose semantics are apparently documented in `:h bufname()`, which says that: _"If `{buf}` is a Number, that buffer number's name is given."_ But I am trying to get the buffer number! 🐔 🔄 🥚
3. Continuing from `:h bufname()`: _"Number zero is the alternate buffer for the current window."_ This is inconsistent with 1.
4. There are many instances of "buffer handle", rather than "buffer number" in `:h api`, and only some of them say "buffer handle (buffer number)" indicating equivalence (the argument is also variously called `{buffer}` or `{buf}`
5. Many api methods that accept a buffer number as an argument treat `0` as an alias for the current buffer's number, which is inconsistent with both 1. and 3.

**For windows:**
1. `nvim_open_win` returns `0` on failure
2. the return value from that same function is documented as a "window handle" which is a confusing term as it does not appear in `:h winid` (windows are already a bit confusing as they have both an ID and a number, unlike buffers, so adding another word to the mix is not a good idea)
3. many other window API methods e.g. `nvim_win_get_config` treat `0` as an alias for the current window's number, which is inconsistent with 1.
4. the function which converts between ID and number, `win_getid`, treats an empty (`nil`) value as signifying the current window number instead
5. `clearmatches` accepts `{win}` which can apparently be either a window ID or a window number (how does that work?)
6. important methods like `nvim_win_is_valid` and `nvim_set_current_win` also use the unclear language "window handle", and use `0` as an alias for the current window's "handle", which means that `vim.api.nvim_win_is_valid(vim.api.nvim_open_win(...))` will confusingly return `true` if opening the new window failed

**For both:**
- `nvim_set_option_value` and `nvim_get_option_value` are rare cases of api functions which specify that the `win` argument in `opts` is a `window-ID`. Unfortunately, this is not really true, as they both happily accept `0` as an alias for the current window. Once again, in combination with `nvim_open_win` failure, this easily leads to setting values on an incorrect window. A similar problem exists for the `buf` argument in `opts`, which also interprets `0` as meaning the current buffer.

This also affects tab page numbers/handles/whatever in a similar way (`0` alias, maybe other things, I don't really use tab pages much).

### What could be done?

Any changes would probably break lots of things, because some of this behaviour is inherited from Vim and many plugins and users will have come to rely on it. Is there any way this could be improved before a version 1 release at least? Here are some ideas:

I can see the convenience in using `0` as an alias for the current window/buffer, especially in the `fn` namespace where methods are intended to be easy to use in frequent cmdline mode calls by users at runtime. However, having either `0` or `nil` mean "current window/buffer" is not ideal. I suggest using `0` for this, everywhere in `fn`, if possible. Having `0` as the alias for the alternate buffer is just confusing though, and could maybe be avoided.

In my opinion, at least the `api` methods, which are neovim-specific, and, well, API rather than intended for frequent use in cmdline mode, should be more consistent and predictable. I would think that there could be a `v:winid_invalid`, `v:win_invalid` and `v:buf_invalid` somewhere that could be cross-referenced in the docs and used consistently in the implementations. Maybe even `v:buf_alt` as the API version of `#`.

There are some easy parts like removing use of "handle" in the API docs and making the behaviour of the option setters/getters explicit. Is any of the rest possible or would it all be too messy at this point?

**Comment by adigitoleo** [comment]
A quick example to demonstrate just one of the ways to get confusing behaviour from the current api:

Run in `nvim -u NORC`:
```lua
:=vim.api.nvim_set_option_value("winfixbuf", true, { win = vim.fn.win_getid(698724) })
```

Then try `:edit`ing any file.

**Comment by bfredl** [comment]
> nvim_create_buf returns 0 on failure

> nvim_open_win returns 0 on failure

These are not really true. API functions throw an error on failure. (an `:echo expression_that_fails()` might show an 0 in addition to the error message, as vim script is weird a f. this has nothing to do with the actual API function).

**Reply by adigitoleo** [reply]
I see, yes that is true. I tried a few simple cases and it's not easy to get 0, so this is less of a concern to me now. I added those points based on the documentation though, which for both of those functions explicitly states that the return value is 0 on error.
